torch_xla/csrc/xla_graph_executor.cpp

```diff
@@ -610,28 +619,85 @@ XLAGraphExecutor::ExecuteComputationWithBarrier(
       << ". Maybe the entry get "
          "kicked out of the LRU cache";
 
+  SyncTensorCollection coll;
+  coll.device = device;
+
   // Create DataPlaceHolder that will get filled in async executions.
   std::vector<xla::Shape>* output_shapes =
       DeviceContextArena::Get()->GetOutputShapesByHash(hash);
   std::vector<torch::lazy::BackendDataPtr> placeholders;
-  placeholders.reserve(output_shapes->size());
-  for (const xla::Shape& shape : *output_shapes) {
-    torch::lazy::BackendDataPtr handle =
-        WrapXlaData(runtime::GetComputationClient()->CreateDataPlaceholder(
-            device.toString(), std::move(shape)));
-    placeholders.push_back(handle);
+  std::vector<XLATensor::ShardingSpecPtr> sharding_specs;
+
+  if (static_cast<XlaDeviceType>(device.type()) == XlaDeviceType::SPMD) {
+    sharding_specs =
+        std::vector<XLATensor::ShardingSpecPtr>(output_shapes->size());
+    static std::unordered_map<torch::lazy::hash_t,
+                              std::vector<XLATensor::ShardingSpecPtr>,
+                              torch::lazy::HashReducer>
+        output_sharding_hash;
+    if (output_sharding_hash.find(hash) == output_sharding_hash.end()) {
+      output_sharding_hash[hash] = ShardingUtil::GetOutputSharding(
+          output_shapes, cachedComputation->computation, device,
+          output_shapes->size());
+    }
+    placeholders =
+        ShardingUtil::CreateShardedPlaceholder(output_sharding_hash[hash]);
+  } else {
+    placeholders.reserve(output_shapes->size());
+    for (const xla::Shape& shape : *output_shapes) {
+      torch::lazy::BackendDataPtr handle =
+          WrapXlaData(runtime::GetComputationClient()->CreateDataPlaceholder(
+              device.toString(), std::move(shape)));
+      placeholders.push_back(handle);
+    }
   }
 
-  SyncTensorCollection coll;
-  coll.device = device;
-  coll.unlocker = DeviceLockerArena::Get()->LockDevices({device});
+  // init this before the lock to save some time.
+  // std::vector<torch::lazy::BackendDataPtr> arguments(graph_inputs.size());
   std::vector<torch::lazy::BackendDataPtr> arguments;
+  arguments.reserve(graph_inputs.size());
+
+  {
+    tsl::profiler::TraceMe activity("DeviceBarrier",
+                                    tsl::profiler::TraceMeLevel::kInfo);
+    TF_VLOG(5) << "Lock device " << device.toString() << "...";
+    coll.unlocker = DeviceLockerArena::Get()->LockDevices({device});
+    TF_VLOG(5) << "Locking device " << device.toString() << " Done!";
+  }
+
   {
     // GetXlaData must be called within a lock region, otherwise it might
     // extract the placeholder inserted by previous execution.
     TORCH_LAZY_TIMED("RunCachedGraphInputData");
+    //   size_t per_thread_work = 200;
```

Comment:
I see a bunch of commented lines here -- maybe we should clean them before mering?